README.md
ZomboTropolis
=============

Join the [Discord Server](https://discord.gg/cnHUgMt)

An **unfinished** zombie survival MMO using Corona and Lua. Play as either a zombie or a human in a fight for survival.  The game is set to replicate a [prey-predator model](http://www.tiem.utk.edu/~gross/bioed/bealsmodules/pred-prey.gph1.gif) of population. 

Human Survivor
==============

*Featuring:*
+ **+40 Different Items**
+ **+40 Skills**

*Multiple Classes:*
+ **Military** - Improved combat abilities 
+ **Research** - Improved healing and gadget capabilities
+ **Engineering** - Improved construction and equipment efficiency  

Play as a human with a simple goal - *survive*. Using many items, gadgets, weapons, and equipment at your disposal you can fend off the undead hordes. Barricade and repair buildings. Search for supplies. Coordinate with the other survivors to stay alive. Humans have the advantage of technology and the ability to interact with their environment to prolong their lifespan, their only true weakness... death.

Undead Zombie
=============

*Featuring:*  
+ **+10 Abilities**
+ **+40 Skills**

*Multiple Classes:*
+ **Brute** - Boosted attacks and lethality
+ **Hunter** - Upgraded movement and tracking utility
+ **Hive** - Dangerous item or building corrosion and infection spread

As a zombie you have one raw need - *food*. And what better food than the fleshy human meatbags that are hiding away in fear? To satiate your hunger, you must work together with your fellow zombie brethren. Use abilities, sheer numbers, and persistence to tear down barricades and reap the soon to be corpses (humans) inside buildings. Zombies do **NOT** fear death. It only slows them down. The only true fear that a zombie has is... starvation.

Time Based Leveling
===================

There is no gathering XP by kills, training, or healing in this game. The only measure is time. The longer your character survives the more abilities and skills they unlock to help with your goals. If you survive long enough, a player can even select a class to unlock for even better skills and perks. Beware though, that the skill system is based on a [parabola](http://www.phas.ubc.ca/~mav/SOP2015/fig3.jpeg). The more skills you have already, the longer it will take to unlock the next one, so pick wisely!

`1.0.0` Release Date
==================

Development is dependent on my free time and level of boredom.  ETA until App Store launch is in 2019.  I will update this timeline with more specific season/month/week dates as I get closer and closer to the finish line.

Credits
=======

A lot of free source assets and libraries were used in the making of ZomboTroplois:

* [RL-Dice](https://timothymtorres.github.io/RL-Dice) by myself
* [middleclass.lua](https://github.com/kikito/middleclass) by kikito
* [Game-Icons.net](http://game-icons.net/) used for skill icons 
* [CoronaSDK](https://coronalabs.com/) used to render the code onto multiple platforms (phone, tablet, etc.)
* [SS13 Sprites](https://github.com/tgstation/tgstation) used for mobs, items, and locations in game
* [Berry](https://github.com/ldurniat/Berry) used to manipulate and display the rooms in game
* [Universal LPC SPritesheet](https://github.com/jrconway3/Universal-LPC-spritesheet) used for the mob and clothing spirtes

Several of these resources are licensed under MIT, CC-BY-SA, or GPL3.

People that helped contribute to this project:

* **TheBeanBurrito** for building the countryside region of the map and game design
* **RedDustBowl** for help in game design and suggestions

Other projects that provided inspiration and ideas for this game:

* [Urbandead](http://www.urbandead.com/) and [Quarantine2019](http://www.quarantine2019.com/) browser based MMO's with the same theme and similiar gameplay.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased Version

## v0.13.0 - ???
### Added
- New human armors: 
	- Biosuit has special effects when worn *depending on the condition of the armor*.  It grants immunity to infection from bites and hunter zombies are unable to track the player.  It also hides the HP status of the player and conceals their location from being revealed by outside zombies with the scent blood skills.  It offers zero protection, is very fragile, and is light weight.  It can be found in labratory buildings and is classified as a research item.  
	- Kevlar is the jack of all trades armor.  It offers moderate protection, has average durability, and average weight.  Additionally it is *the only armor* in the game to have bullet protection.  It can be found in police departments and is classified as a military item.
	- Riotsuit is the king of melee.  It offers high protection from melee attacks, has high durability, but is the heaviest armor available.  It can be found in police departments (very rare) or in the military base and is classified as a military item.
- Pheromone Spray gadget that masks a players scent from any hunter zombies trying to track.  The condition level of the spray determines how much scent is lost.
- New melee weapons: Wrench, Pipe, Rake, Shovel, Axe, Pickaxe, Pitchfork, and Poolstick.
- New ranged weapon: Bow.
- Building machineary now updates durability when used.

### Fixed
- Antidote not properly removing infection status.

### Changed
- Leather Jacket armor is considered bare bones armor.  It offers minimal protection, is fragile, and is the lightest armor in the game.  It can be commonly found in lots of different buildings and is classified as a military item.
- Firesuit armor only protects from acid attacks.  Although acid does no damage to the player, it does destroy their inventory and this armor is the only way to prevent that from happening.  It offers zero protection, is fragile, and is light weight.  It can be found in fire departments and is classified as a engineering item.
- Almost all items have had their damage, durability, accuracy, and weight values changed.
- All the search rates for items.  They now use a google spreadsheet that is exported to a CSV file and loaded into the code from there.  This makes balancing stuff much easier and less of a burden.
- Zombie bite and claw attacks had their damage, accuracy, and skill modifiers all changed.
- Ammo names now should easily reflect what weapon they belong too.
- Renamed the Rifle to be a SMG.
- All items that are found outside (except barricades) will spawn with a high chance of a ruined state.
- Barricades found outside will spawn with a high chance of a worn state.  However barricades found in carparks and junkyards will spawn with a high chance of a pristine state.
- Building integrity values.  Before it was hardcoded at 5 for all buildings.  Now each type of building has it's own value from 3-10 depending on how many resources the building provides. This value represents how much damage a building can sustain and the max length it can be ruined for.  Later when multiblock buildings are added the integrity value will receive a bonus depending on the size. 

### Removed
- Critical hits from all weapons.


## v0.12.0 - 06/14/2019
### Added
- Sprites for all in game items. (some are placeholders)
- Humans interacting with search areas by `Double Tap` results in a single search | `Tap & Hold` results in repeating searches until released.
	- Hidden zombies being revealed by a search.
	- Different search animations for items, junk, and hidden zombies
- Players sprites moving towards the location of the actions they are performing.
- Physics to location scene.  Mobs wander around a location bouncing off walls and physics boundaries.  These physics boundaries seperate attacker/defenders and it depends on which group is in control of the building.  


### Fixed
- Searching causing players to be deleted from location.
- Hidden zombies being displayed in a location.

### Changed
- Equipment, location, item, and mob icons to be placeholder sprites

### Removed
- Axe, machate, and phone items from search chance since the code classes have not been setup.
- Zooming feature for location or world map.  (will be added at a later date)

## v0.11.0 - 11/26/2018
### Added
- Isometric staggered world map.
- Orthogonal location layout that changes depending on player position in world map.
- Buildings have equipment sprites and animations displayed (if powered) for terminal, generator, and transmitters if present.
- Buildings have barricade and door sprites set based on health.
- Locations display mobs inside or outside with their names overhead.  Mobs wander around aimlessly. 
- Dead humans or killed zombies become corpses.  The mob sprite does a 90 degree horizontal rotation to make it appear they are lying down. The sprite also remains motionless.
- Both city and location images can be panned and zoomed within set boundaries.
- Zooming in far enough from the city map will switch to the current room player is in.
- Zooming out far enough from the room will switch to the city world map.

### Changed
- Map tile types to be set based on `Location Tile ID` layer in the `world.tmx` file.  Previously every map tile was set to be a `hospital` type as default, which was used to temporarily get stuff working.  


## v0.10.0 - 01/27/2018
### Added
- `Gadget` and `Syringe` research human skills grant a higher hp threshold when using a syringe on a zombie.
- Scanner item that is used to scan zombies into a database for tracking.  The `Scanner` research human skill boosts the chance of a successful scan attempt.

scenes/action.lua
-----------------------------------------------------------------------------------------
--
-- action.lua
--
-----------------------------------------------------------------------------------------

local composer = require( "composer" )
local widget = require( "widget" )
local action_list = require('code.player.action.list')

local scene = composer.newScene()

-- -----------------------------------------------------------------------------------------------------------------
-- All code outside of the listener functions will only be executed ONCE unless "composer.removeScene()" is called.
-- -----------------------------------------------------------------------------------------------------------------

-- local forward references should go here

local width, height = display.contentWidth, display.contentHeight -- 320x480
local offset, thickness = 100, 3 

-- ScrollView listener
local function scrollListener( event )
   local phase = event.phase
    if ( phase == "began" ) then print( "Scroll view was touched" )
    elseif ( phase == "moved" ) then print( "Scroll view was moved" )
    elseif ( phase == "ended" ) then print( "Scroll view was released" )
    end

    -- In the event a scroll limit is reached...
    if ( event.limitReached ) then
        if ( event.direction == "up" ) then print( "Reached top limit" )
        elseif ( event.direction == "down" ) then print( "Reached bottom limit" )
        elseif ( event.direction == "left" ) then print( "Reached left limit" )
        elseif ( event.direction == "right" ) then print( "Reached right limit" )
        end
    end

    return true
end



local list = {
  location = main_player:getActions('location')
} 

if main_player:isMobType('zombie') then
  list.ability = main_player:getActions('ability')
elseif main_player:isMobType('human') then
  item = main_player:getActions('item')  
end

local menu_w = width - 20
local actionMenu, tabBar, selectedTab, itemListMenu, menu

local function populateItemList()
  menu:insertRow{isCategory=true, params = {name = '[ITEM]', condition = '[CONDITION]', weight = '[SIZE]', canBeTapped=false}}  
  
  for i in ipairs(itemListMenu) do
    local item = itemListMenu[i].item
    menu:insertRow{params = {name = item:getClassName(), condition = item:isConditionVisible(main_player) and item:getConditionState() or '???', weight = item:getWeight(), canBeTapped=true}}
    if itemListMenu[i].collapsed then
      menu:insertRow{params = {name=item:getClassName(), inventory_index=i, discard=true, activate=item:canBeActivated(), canBeTapped=false, item_options=true}}
    end
  end
end

local function onCategoryTap(event)
    local row = event.target
    print("tapped Category", row.id)
for k,v in pairs(itemListMenu) do print(k,v) end
    local id = row.id - 1 -- bc of our top item row category that is omitted
    itemListMenu[id].collapsed = not itemListMenu[id].collapsed
    menu:deleteAllRows()
    populateItemList()
end

local function onRowRender( event )

    -- Get reference to the row group
    local row = event.row

    -- Cache the row "contentWidth" and "contentHeight" because the row bounds can change as children objects are added
    local rowHeight = row.contentHeight
    local rowWidth = row.contentWidth
    
    if row.params.item_options then
      print('')
      print('row.params are:')
      for k,v in pairs(row.params) do print(k,v) end
      
      -- Function to handle button events
      local function handleButtonEvent( event )

          if ( "ended" == event.phase ) then
              print( "Button was pressed and released" )
              local item_name = row.params.name
              print(item_name, 'action cost is - ', main_player:getCost('ap', item_name))
              
              local options = {
                 isModal = true,
                 effect = "fade",
                 time = 400,
                 params = {}
              }             
              
              local mob_type = main_player:getMobType()
              local action_data, params = action_list[mob_type][item_name], options.params
              params.id = action_data.name
              params.inv_id = event.target.id
              params.name = action_data.name
              params.desc = action_data.desc
              params.icon = action_data.icon
              params.cost = main_player:getCost('ap', item_name)
              
              print('')
              print('The params going into options are:')
              for k,v in pairs(options.params) do print(k,v) end
              
              composer.showOverlay( "scenes.action_perform", options)        
              print( "Button was pressed and released" )              
          end
      end

      -- Create the widget
      local use_button = widget.newButton
      {
          left = 20,
          top = 0,
          id = row.params.inventory_index,
          label = "Use",
          shape = 'roundedRect',
          height = 40,
          width = 180,
          onEvent = handleButtonEvent
      }   
      use_button:setFillColor(0, 1, 0)
      row:insert(use_button)
    else
      local rowTitle = display.newText( row, row.params.name, 10, 0, nil, 14 )  -- THIS LINE IS IMPORTANT
      rowTitle:setFillColor( 0 )
      rowTitle.anchorX = 0
      rowTitle.y = rowHeight * 0.5
      
      local condition_margin = -80
      local condition_text_options = {
        parent = row,
        text = row.params.condition,     
        x = 0 + condition_margin,
        y = rowHeight * 0.5,
        width = rowWidth,
        fontSize = 14,
        align = "right",
      }
      local rowTitleCondition = display.newText(condition_text_options)
      rowTitleCondition:setFillColor(0.1, 1, 0.1)  
      rowTitleCondition.anchorX = 0
      
      local weight_margin = -10    
      local weight_text_options = {
        parent = row,
        text = row.params.weight,     
        x = 0 + weight_margin,
        y = rowHeight * 0.5,  
        width = rowWidth,
        fontSize = 14,
        align = "right",  
      }
      local rowTitleWeight = display.newText(weight_text_options)  
      rowTitleWeight:setFillColor(0.1, 0.1, 1)  
      rowTitleWeight.anchorX = 0    
      
      if row.params.canBeTapped then
        local categoryBtn = display.newRect( row, 0, 0, row.width, row.height )
        categoryBtn.anchorX, categoryBtn.anchorY = 0, 0
        categoryBtn:addEventListener ( "tap", onCategoryTap )
        categoryBtn.alpha = 0
        categoryBtn.isHitTestable = true
        categoryBtn.id = row.id
      end
    end
end

local function getMenu(menu_type, menu_data)
  --local menu
  
  if menu_type == 'item' then
    menu = widget.newTableView
    {
        top = 80,
        left = 10,
        width = menu_w,
        height = height - 150,
        onRowRender = onRowRender,
      --onRowTouch = onRowTouch,
        isBounceEnabled = false,
        listener = scrollListener
    }

    
    itemListMenu = {}   
    for i, item in ipairs(menu_data) do
      itemListMenu[#itemListMenu+1] = {item = item, collapsed=false}
    end    
    populateItemList()
  else    
    local num, row = 0, 0
    local padding, offset = 10, 20
    local divider_w, divider_h = 20, 15
    --  local icon_size_w, icon_size_h = 90, 40 
    local icon_radius = 40
    local icon_size = icon_radius*2    
    
    menu = widget.newScrollView
    {
        top = 80,
        left = 10,
        width = menu_w,
        height = height - 150,
        scrollWidth = 600,
        scrollHeight = 800,
        listener = scrollListener
    }  
    
    
    local mob_type = main_player:getMobType()
    
    local options = {
       isModal = true,
       effect = "fade",
       time = 400,
       params = {
          sampleVar = "my sample variable"
       }
    }        
    
    for _, action in ipairs(menu_data) do
      if row == 0 then
        num = num + 1
      end       
    
      local buttonEvent = function(event)
        if ("ended" == event.phase ) then
print(event.target.id, 'action cost is - ', main_player:getCost('ap', event.target.id))
          local action_data, params = action_list[mob_type][event.target.id], options.params
          params.id = event.target.id
          params.name = action_data.name
          params.desc = action_data.desc
          params.icon = action_data.icon
          params.cost = main_player:getCost('ap', event.target.id)
                
          composer.showOverlay( "scenes.action_perform", options )        
          print( "Button was pressed and released" )
        end
      end     
    
      local button = widget.newButton
      {
          left = padding + icon_size*(row) + divider_w*(row),
          top = offset + icon_size*num + divider_h*num,
          id = action,
      --  label = skill_data.name,
          onEvent = buttonEvent,
          shape = 'circle',
          radius = icon_radius,
          fillColor = { default={ 1, 0, 0, 1 }, over={ 1, 0.1, 0.7, 0.4 } },
          strokeColor = { default={ 1, 0.4, 0, 1 }, over={ 0.8, 0.8, 1, 1 } },
          strokeWidth = 3      
      }      
      
      -- Change the button's label text
      button:setLabel(action)              
    --  skill_data.tier
      menu:insert( button )  
      
      if row == 0 then
        row = row + 1
      elseif row == 2 then
        row = 0
      else
        row = row + 1
      end      
    end
  end
  return menu
end 

local function redoMenu(menu_type)
  list[menu_type] = main_player:getActions(menu_type)  
  actionMenu:removeSelf()
  actionMenu = nil
  return getMenu(menu_type, list[menu_type])
end
-- -------------------------------------------------------------------------------


-- "scene:create()"
function scene:create( event )
  local sceneGroup = self.view
  local drawStatBar, stat_bar

  -- Initialize the scene here.
  -- Example: add display objects to "sceneGroup", add touch listeners, etc.
 
  -- Function to handle button events
  local function handleTabBarEvent( event )
      print('event.target._id ', event.target._id )  --reference to button's 'id' parameter
    selectedTab = event.target._id
    actionMenu = redoMenu(event.target._id)
    sceneGroup:insert(actionMenu)
  end   
 
  -- Configure the tab buttons to appear within the bar
  local tabButtons = {
      {
          label = "Location",
          id = "location",
          selected = true,
          onPress = handleTabBarEvent
      },
  }
  
  if main_player:isMobType('zombie') then
      tabButtons[#tabButtons+1] = {
          label = "Ability",
          id = "ability",
          onPress = handleTabBarEvent
      }    
  elseif main_player:isMobType('human') then
      tabButtons[#tabButtons+1] = {  
          label = "Item",
          id = "item",
          onPress = handleTabBarEvent
      }        
  end  
  
  selectedTab = 'location'
  
  -- Create the widget
  local tabBar = widget.newTabBar
  {
      top = 20,
      left = 10,
      width = menu_w,
      height = 60,      
      buttons = tabButtons
  }   

  -- Create a image and insert it into the scroll view
  --local background = display.newImageRect( "assets/scrollimage.png", 768, 1024 )
  --scrollView:insert( background )
  
  local stat_bar_h, stat_bar_w = 12, height/9 + offset
  
  function drawStatBar()
    stat_bar = display.newGroup()
    local background_bar = display.newRect(stat_bar_w+offset, stat_bar_h*0.5 + thickness, stat_bar_w, stat_bar_h)
    background_bar.strokeWidth = thickness
    background_bar:setFillColor( 0.5 )
    background_bar:setStrokeColor( 1, 0, 0 )
    stat_bar:insert(background_bar)

    local stat_text
    if main_player:isMobType('human') then      
      local ip, max_ip = main_player:getStat('ip'), main_player:getStat('ip', 'max')
      stat_text = display.newText('IP ['..ip..'/'..max_ip..']', stat_bar_w+offset, stat_bar_h*0.5 + thickness, native.systemFontBold, 10 )
    elseif main_player:isMobType('zombie') then
      local ep, max_ep = main_player:getStat('ep'), main_player:getStat('ep', 'max')
      local decay, max_decay = main_player.condition.decay:getTime(), 1023
      stat_text = display.newText('EP ['..ep..'/'..max_ep..']  DECAY ['..decay..'/'..max_decay..']', stat_bar_w+offset, stat_bar_h*0.5 + thickness, native.systemFontBold, 10 )      
    end
    stat_text:setFillColor( 0, 0, 0 )
    stat_bar:insert(stat_text)    
    sceneGroup:insert(stat_bar)
  end  
  
  actionMenu = getMenu('location', list.location)
  drawStatBar() 
  sceneGroup:insert(tabBar)
  sceneGroup:insert(actionMenu) 
end


-- "scene:show()"
function scene:show( event )

    local sceneGroup = self.view
    local phase = event.phase

    if ( phase == "will" ) then
        -- Called when the scene is still off screen (but is about to come on screen).
        --[[
  -- Function to handle button events
  local function handleTabBarEvent( event )
      print( event.target._id )  --reference to button's 'id' parameter
    selectedTab = event.target._id
    actionMenu = redoMenu(event.target._id, list[event.target._id])
    sceneGroup:insert(actionMenu)
  end         
        --]]
print()
print('action scene phase == "will"')
print('selectedTab is - ', selectedTab)
for k,v in pairs(list[selectedTab]) do print(k,v) end
print()
        actionMenu = redoMenu(selectedTab)
        sceneGroup:insert(actionMenu)        
    elseif ( phase == "did" ) then
        -- Called when the scene is now on screen.
        -- Insert code here to make the scene come alive.
        -- Example: start timers, begin animation, play audio, etc.
    end
end


-- "scene:hide()"
function scene:hide( event )

    local sceneGroup = self.view
    local phase = event.phase

    if ( phase == "will" ) then
        -- Called when the scene is on screen (but is about to go off screen).
        -- Insert code here to "pause" the scene.
        -- Example: stop timers, stop animation, stop audio, etc.
    elseif ( phase == "did" ) then
        -- Called immediately after scene goes off screen.
    end
end


-- "scene:destroy()"
function scene:destroy( event )

    local sceneGroup = self.view

    -- Called prior to the removal of scene's view ("sceneGroup").
    -- Insert code here to clean up the scene.
    -- Example: remove display objects, save state, etc.
end


-- -------------------------------------------------------------------------------

-- Listener setup
scene:addEventListener( "create", scene )
scene:addEventListener( "show", scene )
scene:addEventListener( "hide", scene )
scene:addEventListener( "destroy", scene )

-- -------------------------------------------------------------------------------

return scene